[STEP 1] aCafela coffee, jerry

aCafela-coffee opened the pull request:
@ryan-son 
안녕하세요 라이언!
순서도는 README 참고 부탁드리구요. STEP 1 까지 진행하고 PR 요청드립니다. 감사합니다.

## 고민되었던 점

- checkTheResult 함수의 사용 시점에 문장처럼 보여지기 위해서 함수의 전달인자에 별칭을 추가하였습니다.
    
    ```swift
    let (strikeCount, ballCount) = checkTheResult(for: playNumbers)
    ```
    
- checkTheResult 함수 내에서 일치하는 수의 개수를 구할 때 들여쓰기 2번이하 사용해야 해서 중첩 for문을 쓰는 대신 contains를 사용하였습니다.
    
    ```swift
    for inputNumber in inputNumbers {
            if randomNumbers.contains(inputNumber) {
                sameNumberCount += 1
            }
        }
    ```
    
- 중복이 허용되지 않는 배열을, Set이 아닌 Array로 만든 이유:
    스트라이크 판정을 위해서 순서가 있는 배열을 사용해야 합니다. Set은 순서가 없기 때문에, 지금은 적절하지 않습니다.
    

## 조언 얻고 싶은것

함수명 네이밍을 하면서 고민이 많았는데요. 

> Include all the words needed to avoid ambiguity
> 

[API Design Guidelines](https://swift.org/documentation/api-design-guidelines) 문서의 이 문장을 인용해서 3개의 숫자를 생성한다라는 기능의 명확히 하기위해 `generate3RandomNumbers` 라는 이름을 지었어요. 함수명 중간에 숫자가 들어가도 되는지 궁금합니다! 

안된다면 다른 방식이 있는지 궁금하네요

aCafela-coffee committed:
Add ios-number-baseball.drawio

aCafela-coffee committed:
chore: Delete '.drawio' file

aCafela-coffee committed:
docs: Add ios-number-baseball.drawio.png

aCafela-coffee committed:
docs: Add image to README.md

aCafela-coffee committed:
feat: 전역변수 추가

llghdud921 committed:
feat: 랜덤숫자 생성 함수 추가

aCafela-coffee committed:
feat: 사용자에게 입력받는 함수를 추가

llghdud921 committed:
feat : 게임 결과를 출력하는 함수를 추가

aCafela-coffee committed:
feat: 게임 시작 함수 추가, STEP 1 완성

aCafela-coffee committed:
chore: 주석 삭제

llghdud921 committed:
style: checkTheResult함수의 전달인자 별칭 추가

ryan-son reviewed (changes requested):
아카페라, 제리! STEP 1 수행하시느라 고생 많으셨어요. 👏

## Note
프로그래밍에는 정답이 없습니다. 따라서 리뷰어가 이야기하는 내용은 정답이 아니므로 꼭 따라야할 필요는 없습니다. 
일반적인 상황에서 좋은 코드와 나쁜 코드를 정의는 할 수 있겠지만, 인력/시간/비용과 같은 가용한 자원과 환경을 고려하면 일반적으로 좋은 코드도 나쁜 코드가 될 수 있으며 그 반대의 경우도 충분히 가능합니다.
따라서 제가 제리와 아카페라의 코드에 대해 일부 지적을 하였다고 해서 해당 코드가 나쁜 코드라는 의미도, 하물며 틀렸다는 의미도 아닙니다. 다만 리뷰이와 리뷰어가 논의해서 최선의 방향으로 진행시켜 나가고자하는 의도일 뿐입니다. 그러니 코드를 작성하실 때는 구성한 논리가 타당하고 합리적이라고 받아 들여질 수 있는 정도로 판단하셔서 작성하시고 그 부분을 저와 논의하여 진행해나가시면 됩니다.

리뷰 중 이해가 안되는 부분이 있으시면 디스코드로 말씀해주시면 확인하는대로 답변드리겠습니다.

## Good Point
제가 느꼈던 좋은 점들이에요.
1. 리뷰를 위한 첫 PR일텐데도 불구하고 커밋 메시지 규칙이 잘 정리되어 각 커밋에서 누가 어떤 작업을 수행했는지 쉽게 파악할 수 있었어요. 
2. 커밋이 작은 단위로 나뉘어 문제가 생겼을 때에도 큰 리스크 없이 revert를 통해 이전 기록으로 돌아갈 수 있을 것 같다고 판단되었어요.
3. 번갈아가며 커밋하신 기록을 보니 페어 프로그래밍을 수행해서 제리와 아카페라 두 분 모두 코드에 대해 충분한 논의를 하였고 작성하신 내용을 잘 이해하고 있을 것 같다고 생각했어요.
4. 정성스레 준비하신 순서도 덕에 이해가 잘 되었고, 소스 코드를 작성하시기 전에 로직에 대해 충분히 논의하고 시작하셨다는 인상을 받았어요.
5. 코드를 통해 줄바꿈을 사용하는 의도가 전해져서 의미 단위를 파악하기 용이했어요.

## Comment Review
> ## 고민되었던 점
> 
> - checkTheResult 함수의 사용 시점에 문장처럼 보여지기 위해서 함수의 전달인자에 별칭을 추가하였습니다.

덕분에 해당 함수의 가독성이 좋았어요. 앞으로도 [API Design Guidelines의 Argument labels 명명규칙](https://swift.org/documentation/api-design-guidelines/#argument-labels)에 따라 적절히 적용해보시길 바라요~!

> - checkTheResult 함수 내에서 일치하는 수의 개수를 구할 때 들여쓰기 2번이하 사용해야 해서 중첩 for문을 쓰는 대신 contains를 사용하였습니다.

관점에 따라 작성하신 내용도 `for-if`로 2단계 코드 깊이를 가지고 있다고 볼 수도 있어요. 생각해볼만한 포인트를 개별 리뷰에 남겨두었으니 참고하세요~

> - 중복이 허용되지 않는 배열을, Set이 아닌 Array로 만든 이유:
>     스트라이크 판정을 위해서 순서가 있는 배열을 사용해야 합니다. Set은 순서가 없기 때문에, 지금은 적절하지 않습니다.
- 이미 알고계시듯이 `Set` 자료형을 사용하면 해당 변수에 새로운 요소를 추가할 때 중복 여부를 판단하지 않아도 된다는 장점이 있어요.
- [`Array` 이니셜라이저](https://developer.apple.com/documentation/swift/array/2907560-init)를 이용하면 `Set`을 `Array`로 변환할 수 있으니 `Set`자료형 사용을 검토해볼 수 있지 않을까요?

어떤 방식을 취할 것이냐는 취향의 차이일 것 같네요.
1. 현행대로 최초 변수부터 Array로 구성하되 중복되는 요소를 판단하여 append 작업을 수행한다.
2. `Set`을 이용하여 중복되는 요소를 판단하지 않고 insert 작업을 수행한 후 Array 자료형으로 변환한다.

> 
> ## 조언 얻고 싶은것
> 
> 함수명 네이밍을 하면서 고민이 많았는데요. 
> 
> > Include all the words needed to avoid ambiguity
> 
> [API Design Guidelines](https://swift.org/documentation/api-design-guidelines) 문서의 이 문장을 인용해서 3개의 숫자를 생성한다라는 기능의 명확히 하기위해 `generate3RandomNumbers` 라는 이름을 지었어요. 함수명 중간에 숫자가 들어가도 되는지 궁금합니다! 
> 
> 안된다면 다른 방식이 있는지 궁금하네요

이미 작성해보셔서 아시겠지만 물론 문법적으로 허용하고 있으며 타당한 이유가 있으면 사용해도 됩니다. 이 주제는 개별 리뷰에 상세한 내용을 남겨두었으니 참고하시길 바라요~

aCafela-coffee committed:
refactor: 무작위 숫자를 생성하는 함수를 수정

- generate3RandomNumbers 함수의 이름을 generateRandomNumbers(count:)로 변경함
- numbersCount 전역변수 추가

llghdud921 committed:
refactor: 전역변수 추가

numberRange 전역변수 추가
띄어쓰기 수정
필요없는 괄호 제거

aCafela-coffee commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,64 @@
+var randomNumbers: [Int] = []
+var remainChances: Int = 9
 
+func generate3RandomNumbers() -> [Int] {
```
-  `generateRandomNumbers(count:)` 함수명을 변경하고 전달인자를 추가해서, 생성할 숫자의 갯수를 입력받도록 변경했습니다.
- `let numberRange: ClosedRange<Int> = 1...9` 전역변수에서 숫자 범위를 지정하도록 변경했습니다.
```swift
func generateRandomNumbers(count: Int) -> [Int] {
    var numbers: [Int] = []
    
    while numbers.count < count {
        let randomNumber:Int = Int.random(in: numberRange)
        
        if numbers.contains(randomNumber) == false {
            numbers.append(randomNumber)
        }
    }
    
    return numbers
}
```

aCafela-coffee commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,64 @@
+var remainChances: Int = 9
 
+func generate3RandomNumbers() -> [Int] {
+    var numbers:[Int] = []
```
수정했습니다.

aCafela-coffee commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,64 @@
 
 import Foundation
```
당장은 import하지 않아도 문제가 없네요. 나중에 필요하면 다시 추가하도록 하겠습니다.

aCafela-coffee committed:
chore: import Foundation 삭제

llghdud921 committed:
refactor: 중복하지 않는 숫자를 생성하는 로직 변경

llghdud921 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,64 @@
+        
+        if numbers.contains(randomNumber) == false {
+            numbers.append(randomNumber)
+        }
```
말씀하신대로 가독성을 위해 `==` 비교연산자를 사용했습니다.

llghdud921 commented:
>이미 알고계시듯이 Set 자료형을 사용하면 해당 변수에 새로운 요소를 추가할 때 중복 여부를 판단하지 않아도 된다는 장점이 있어요.
>Array 이니셜라이저를 이용하면 Set을 Array로 변환할 수 있으니 Set자료형 사용을 검토해볼 수 있지 않을까요?
 
```swift
func generateRandomNumbers(count: Int) -> [Int] {
    var numbers: Set<Int> = []
    
    while numbers.count < count {
        let randomNumber:Int = Int.random(in: numberRange)
        
        numbers.insert(randomNumber)
    }
    
    return Array(numbers)
}
```
아카페라와 논의한 결과, Set을 사용해도 된다고 생각해서 로직을 수정하였습니다.

aCafela-coffee committed:
refactor: checkTheResult(for:) 함수의 리턴값에 이름 추가

llghdud921 committed:
refactor: if문을 삼항연산자로 변경

llghdud921 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,64 @@
+        if randomNumbers.contains(inputNumber) {
+            sameNumberCount += 1
+        }
+    }
```
삼항연산자로 로직을 변경했습니다.

aCafela-coffee committed:
refactor: checkTheResult 함수에서 조건문을 where절로 대체

llghdud921 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,64 @@
+        }
+    }
+
+    return (strikeCount, sameNumberCount - strikeCount)
```
ryan이 알려주신 방법이 연산과정을 줄일 수 있어서 코드 수정했습니다!

llghdud921 committed:
refactor: strike, ball 개수 구하는 로직 변경

checkTheResult 함수 내 변수 수정

aCafela-coffee committed:
refactor: 게임시작 함수의 이름을 변경

aCafela-coffee commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,64 @@
+    return (strikeCount, sameNumberCount - strikeCount)
+}
+
+func gameStart() {
```
`gameStart()` 함수의 이름을 `startGame()` 으로 변경합니다.

aCafela-coffee commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,64 @@
 import Foundation
 
-print("Hello, World!")
+var randomNumbers: [Int] = []
+var remainChances: Int = 9
```
`remainingChance` 이 이름이 더 좋군요. 수정합니다.

aCafela-coffee committed:
refactor: remainChances변수의 이름을 remainingChance로 변경

llghdud921 committed:
refactor: 임의의수 출력용 상수를 선언, 띄어쓰기 수정

aCafela-coffee committed:
refactor: 승리 판정 함수 분리

ryan-son commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,64 @@
+        if randomNumbers[index] == inputNumbers[index] {
+            strikeCount += 1
+        }
+    }
```
코드를 처음 읽어 스트라이크, 볼 판정 로직에 대한 이해를 잘 못하는 사람은 `randomNumbers[index] == inputNumbers[index]`가 무엇을 위한 조건인지 판단하기 어려울 수 있어요. 그래서 아래와 같이 작성해보면 어떨까하는 의견입니다!

```swift
for index in 0..<numbersCount {
    let isStrike: Bool = randomNumbers[index] == inputNumbers[index]

    if isStrike {
      // do something
    } else ...
} 
```

ryan-son commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,64 @@
+    
+    while remainChances > 0 {
+        let playNumbers: [Int] = generate3RandomNumbers()
+        print("임의의 수 : \(playNumbers.map{ String($0) }.joined(separator: " "))")
```
잘 이해하고 계시니 상황에 맞게 축약형 표현을 적절하게 활용하시면 되겠습니다. 당분간은 클로저 축약 표현을 사용하실 때 챌린지를 받으실 수도 있으실텐데, 이 정도의 이해를 가지고 있다라고 말씀해주시면 설득하시기 편하실 거에요. 하지만 리뷰어님의 성향에 따라 다른 의견을 가지고 계실 수 있으니 논의하셔서 작성 스타일을 맞추면 되겠습니다!

- 마크다운 인라인 코드 문법 때문에 한 줄로 작성하셨겠지만, 클로저의 줄바꿈 코딩 컨벤션도 기억해두시면 좋아요~
```swift
playNumbers.map { (number: Int) -> String in
    return String(number) 
}
```

- > 함수의 유일한 인자인 경우 소괄호를 생략 가능

말씀해주신 내용은 함수의 매개변수 전달에 사용되는 소괄호 전체를 제거할 수 있는 조건이네요. 클로저는 함수 또는 메서드 마지막 매개변수에 해당할 때 클로저를 감싸는 소괄호를 생략할 수 있습니다. 3 번 링크의 Trailing closure (후행 클로저) 내용을 참고하세요.

- 추가로 참고할만한 내용은 함수 또는 메서드가 복수의 클로저를 매개변수로 받을 때 스타일링 관련 (줄바꿈 규칙과 후행 클로저를 적용 여부) 내용이 있겠네요.

정답은 아니지만 [Raywenderlich Swift Style Guide](https://github.com/raywenderlich/swift-style-guide#closure-expressions)의 논리와 예시를 보시고 왜 이렇게 사용하는 것을 선호하는지 알아보는 것도 재미있을거에요.

ryan-son commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -3,8 +3,124 @@
-print("Hello, World!")
+var randomNumbers: [Int] = []
+var remainingChance: Int = 9
+let numbersCount: Int = 3
+let numberRange: ClosedRange<Int> = 1...9
```
현재 네이밍에 만족하신다면 그대로 유지하셔도 좋고, 이름만으로는 `number`에 대한 설명이 부족하다고 느끼신다면 이를 설명하는 문구를 붙여도 좋을 것 같아요. 저는 `inputNumberRange` 같은 이름이 떠오르는데, 더 좋은 이름이 있다면 변경해볼까요?

llghdud921 replied:
inputNumberRange 좋은것같아요! 수정완료했습니다.

ryan-son commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -4,7 +4,68 @@
+    var strikeCount: Int = 0
+    var ballCount: Int = 0
+    
+    for index in 0..<numbersCount{
+        if randomNumbers[index] == inputNumbers[index] {
+            strikeCount += 1
+        } else if randomNumbers.contains(inputNumbers[index]) {
+            ballCount += 1
+        }
+    }
```
이전 코멘트에서는 아래 내용을 제안 드린 것이에요. 현행을 유지할지, 아래 내용을 적용할지 여부는 논의하셔서 결정하시면 됩니다!

반복문의 코드블럭 공백 형식도 신경써주세요~!

```suggestion
    for index in 0..<numbersCount {
        let isStrike: Bool = randomNumbers[index] == inputNumbers[index]
        let isBall: Bool = randomNumbers.contains(inputNumbers[index])
    
        if isStrike {
            strikeCount += 1
        } else if isBall {
            ballCount += 1
        }
    }
```

ryan-son commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -4,7 +4,68 @@
+    return Array(numbers)
+}
 
+func checkTheResult(for inputNumbers: [Int]) -> (strikeCount: Int, ballCount: Int) {
```
함수 이름에 관사를 유지하시기로 한 이유가 있을까요?

ryan-son commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -4,7 +4,68 @@
+    while remainingChance > 0 {
+        let playNumbers: [Int] = generateRandomNumbers(count: numbersCount)
+        let playNumbersForPrint: String = playNumbers.map { String($0) }.joined(separator: " ")
+        let result = checkTheResult(for: playNumbers)
```
`result`의 의미를 명확히 하기 위한 문구가 하나 더 있으면 좋겠어요. `playNumbers`가 있으니 `playResult`도 괜찮겠네요. 이렇게 이름을 바꾼다면 일관성을 유지하기 위해 함수 이름도 `checkPlayResult(for:)`로 변경할 수 있을 것 같아요.

ryan-son commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -4,7 +4,68 @@
+        }
+        
+        remainingChance -= 1
+        print("\(result.strikeCount) 스트라이크, \(result.ballCount) 볼")
+        print("남은 기회 : \(remainingChance)")
```
이 친구들도 분리하면 하나의 책임을 가진 함수로 만들 수 있을 것 같네요!

ryan-son commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -4,7 +4,68 @@
+        print("남은 기회 : \(remainingChance)")
+    }
+    
+    printWin()
```
승리 출력이라는 의미도 좋지만 개인적으로는 승자를 출력한다라는 의미가 더 어울려 보이는데, 어떻게 생각하시나요?

ryan-son commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -4,7 +4,68 @@
+}
+
+
+// MARK: Game start
```
Mark 주석을 사용해보셨군요! 일반 마크주석 이후에 대시(-)를 추가하시면 소스코드에 구분선을 표시하실 수 있어요. `// Mark: -`

두 분이 논의하셔서 지금 상황에서 필요할지, 언제 구분선이 필요할지 생각해보시길 바랍니다! 이것도 마찬가지로 논의하기 나름이라 정답이 없어요!

ryan-son reviewed (approved):
리뷰 내용을 신속하게 반영하셨네요. 고생 많으셨습니다! 👏

이번에는 스타일링 이외에는 이견이 없어 코멘트만 남기고 바로 머지하도록 하겠습니다.
이번에 리뷰드린 내용들은 다음 스텝에서 반영할지 여부를 판단해주세요~!
